File: 3.longest-substring-without-repeating-characters.cpp
/*
 * @lc app=leetcode id=3 lang=cpp
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int map[256] = {0};
        int l = 0;
        int n = s.size();
        int result = 0;
        for (int i = 0; i < n; i++) {
            if (map[s[i]] > 0 && map[s[i]] >= l) {
                l = map[s[i]];
            }
            map[s[i]] = i + 1;
            result = max(result, i - l + 1);
        }

        return result;
    }
};
// @lc code=end

File: 3.longest-substring-without-repeating-characters.java
/*
 * @lc app=leetcode id=3 lang=java
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int[] dict = new int[256];
        int l = 0, r = 0, max = 0, n = s.length();
        while (r < n) {
            char c = s.charAt(r++);
            if (dict[c] == 0) {
                dict[c] = r;
                max = Math.max(max, r - l);
                continue;
            }
            for (; l < dict[c]; l++) {
                dict[s.charAt(l)] = 0;
            }
            dict[c] = r;
        }
        return max;
    }
}
// @lc code=end

